airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java

```diff
@@ -358,25 +358,43 @@ public WebBackendConnectionRead webBackendCreateConnection(final WebBackendConne
+   * As a convenience to the front-end, this endpoint also creates new operations present in the
+   * request, and bundles those newly-created operationIds into the connection update.
+   */
+  public WebBackendConnectionRead webBackendUpdateConnection(final WebBackendConnectionUpdate webBackendConnectionPatch)
       throws ConfigNotFoundException, IOException, JsonValidationException {
-    final List<UUID> operationIds = updateOperations(webBackendConnectionUpdate);
-    final ConnectionUpdate connectionUpdate = toConnectionUpdate(webBackendConnectionUpdate, operationIds);
-    final UUID connectionId = webBackendConnectionUpdate.getConnectionId();
-    final ConfiguredAirbyteCatalog existingConfiguredCatalog =
+
+    final UUID connectionId = webBackendConnectionPatch.getConnectionId();
+    ConnectionRead connectionRead = connectionsHandler.getConnection(connectionId);
+
+    // before doing any updates, fetch the existing catalog so that it can be diffed
+    // with the final catalog to determine which streams might need to be reset.
+    final ConfiguredAirbyteCatalog oldConfiguredCatalog =
         configRepository.getConfiguredCatalogForConnection(connectionId);
-    ConnectionRead connectionRead;
-    connectionRead = connectionsHandler.updateConnection(connectionUpdate);
 
-    final Boolean skipReset = webBackendConnectionUpdate.getSkipReset() != null ? webBackendConnectionUpdate.getSkipReset() : false;
+    final List<UUID> operationIds = updateOperations(connectionRead, webBackendConnectionPatch);
+
+    // pass in operationIds because the patch object doesn't include operationIds that were just created
+    // above.
+    final ConnectionUpdate connectionPatch = toConnectionPatch(webBackendConnectionPatch, operationIds);
+
+    // persist the update and set the connectionRead to the updated form.
+    connectionRead = connectionsHandler.updateConnection(connectionPatch);
+
+    // now that the connection is fully updated, check for a diff between the
+    // old catalog and the updated catalog to see if any streams need to be reset.
+    final Boolean skipReset = webBackendConnectionPatch.getSkipReset() != null ? webBackendConnectionPatch.getSkipReset() : false;
     if (!skipReset) {
-      final AirbyteCatalog apiExistingCatalog = CatalogConverter.toApi(existingConfiguredCatalog);
-      final AirbyteCatalog newAirbyteCatalog = webBackendConnectionUpdate.getSyncCatalog();
-      newAirbyteCatalog
-          .setStreams(newAirbyteCatalog.getStreams().stream().filter(streamAndConfig -> streamAndConfig.getConfig().getSelected()).toList());
-      final CatalogDiff catalogDiff = connectionsHandler.getDiff(apiExistingCatalog, newAirbyteCatalog);
+      final AirbyteCatalog apiExistingCatalog = CatalogConverter.toApi(oldConfiguredCatalog);
+      final AirbyteCatalog upToDateAirbyteCatalog = connectionRead.getSyncCatalog();
+      final CatalogDiff catalogDiff = connectionsHandler.getDiff(apiExistingCatalog, upToDateAirbyteCatalog);
```

Comment:
thinking out loud: the section on resetting is long, what about pulling it out into a private method named `resetStreamsIfNeeded` or something along those lines.